feat: annotator component #227

mturoci opened the pull request:
Closes #227

shanaka-rajitha commented:
@lo5 @mturoci  I have updated few design samples based on the initial completed design with minor variations. [Design URL](https://xd.adobe.com/view/26f90005-a225-4e48-af10-e6e082222bd9-73aa/)

![text annotator layout](https://user-images.githubusercontent.com/12246331/116042581-303f0200-a68c-11eb-88bb-6437194df15a.png)

mturoci commented:
Looks good @shanaka-rajitha! @lo5 I suppose we are going with the second (right side) option that resembles our card more.

lo5 commented:
@mturoci - yes

I like the muted colors in the second design. I liked the position of the 'x' in Martin's implementation - allows us to show-on-hover. The inline 'x' in the second design adds to visual noise. It would be good if the highlighted text were inconspicuous.

shanaka-rajitha commented:
@lo5 @mturoci, Iv updated the annotator design based on feedback and comments. However I feel we can add more line height to allow some breathing space. With current themeing line height is restricted to the font size used.

[Design URL](https://xd.adobe.com/view/26f90005-a225-4e48-af10-e6e082222bd9-73aa/)

![text annotator layout](https://user-images.githubusercontent.com/12246331/116555973-93de5f00-a91a-11eb-808e-f7ed2cd6fa40.png)

lo5 commented:
![leftright](https://user-images.githubusercontent.com/6334572/116582391-2d315500-a8ca-11eb-8e42-013e146b66d5.png)

If we go with the right-aligned 'x' button strategy, the 'x' will end up in the lower-middle [1] if the text wraps.

[2] is not affected by word-wrap.

Also the 'x' buttons in the design look extra-small - the 'x' is not legible, and requires more precision to click on.

shanaka-rajitha commented:
@mturoci I have updated the designs based on the feedback. Following items were improved 

- Close icon size increased from `9px` to `16px`.
- Highlighted text will have a different font color, this is to allow font visibility based on the background.
- Font weight changed to `300` for improved clarity and breathing space.

mturoci commented:
Changed accroding to design:

https://user-images.githubusercontent.com/64769322/116684833-c4250c80-a9b1-11eb-927a-0be5ad89ef8a.mov

* Close icon size increased to `18px` - XD seems to render icons weirdly, it should be font based, but it seems it has a container (black circle) and text with it.
* Nit: Maybe we could be consistent in border-radiuses - same for header and highlighted text.
* Changed UX of removing to only appear on hover.

shanaka-rajitha commented:
Looks good @mturoci 💯

- we can make the shapes of the tags and highlight section similar, refer attached image which shows first design with 4px radius and the second with no radius.
- the shadow on tags is it so show selected state.

<img width="1258" alt="Screenshot 2021-04-30 at 18 17 04" src="https://user-images.githubusercontent.com/12246331/116697410-c6eb2600-a9e0-11eb-9bda-d6e5762e756b.png">

lo5 commented:
Observations: 
- It's odd that the word-spacing increases when you highlight text.
- The selected-state for tags needs design @shanaka-rajitha (not sure if the drop shadow is the best mechanism here - any alternatives?)

mturoci commented:
> It's odd that the word-spacing increases when you highlight text.

caused by padding - the same what prodigy does to make a better distinction beween highlighted and non-highlighted text. We can go with just background color change ofc.

lo5 commented:
> caused by padding

Padding around the highlighted section is fine. Padding around individual words within the section is not.

mturoci commented:
Padding fixed, border radius changed to `2px` - same as Fluent uses.

https://user-images.githubusercontent.com/64769322/116732919-b0969780-a9eb-11eb-836b-ab5dde325618.mov

shanaka-rajitha commented:
@lo5 I have shared 2 samples for selected-state of tags which is very subtle. Since the tag colors vary from each other having a color shade for active tag would not highlight the purpose. 

<img width="1110" alt="selected state of tags UI" src="https://user-images.githubusercontent.com/12246331/116913872-a1bc1900-ac67-11eb-80d8-289101b47322.png">

lo5 commented:
How about:
1. Gray out (or fade out?) other tags when a tag is active.
2. Make hover state = normal state for grayed-out tags.

shanaka-rajitha commented:
@lo5 fading out is a good approach

1. On active state other tags will fade in color or be greyed.
2. When hovered over a a faded tag it will show the default color.

<img width="740" alt="Tag active state-v2" src="https://user-images.githubusercontent.com/12246331/117010548-881fdd80-ad0a-11eb-944e-1aadfba3224c.png">
<img width="745" alt="Tag active state-v1" src="https://user-images.githubusercontent.com/12246331/117010565-8bb36480-ad0a-11eb-9da5-243dfaa1dc33.png">

lo5 commented:
![2021-05-13_10-39](https://user-images.githubusercontent.com/6334572/118164148-9db29880-b3d7-11eb-85b5-2011ff63a278.png)

- Per design, height of tags [1] needs to be smaller, or match the height of the tagged text. Currently it's taller than a button [2].
- [3] [4] compare left vs right padding when you select text.
- [5] [6] looks like border-radius is missing?
- [7] Fade/dim needs to be more pronounced (see design).

mturoci commented:
Redesigned according the spec.
> Per design, height of tags [1] needs to be smaller, or match the height of the tagged text. Currently it's taller than a button [2].

Not sure where this came from. In the design spec, the tag height is `32px`, content line height `25px` and submit button `42px` (Fluent has button height `32px`). I have fixed it according to design spec with mentioned values with the exception of Fluent button.

> [3] [4] compare left vs right padding when you select text.

Fixed.

> [5] [6] looks like border-radius is missing?

Fixed.

> [7] Fade/dim needs to be more pronounced (see design).

IMO the proposed design with gray color is not ideal. It uses `neutralLight` color, which might not be a good fit with different theme (right now it looks good because it resembles disabled state, but that may not be the case with other themes). Secondly, even the current contrast between background and button text is pretty low, making it hard to read. I consider previous opacity reduction better.

Also some design spacing values are a bit strange:
* tagged text - padding: left `11px`, right `10px`, top `1px`, bottom `2px`
* tag padding: top `6px`, bottom `5px`

Result:
![image](https://user-images.githubusercontent.com/64769322/118248091-331b5e80-b4a4-11eb-9c7f-eca0c8691c0b.png)

mturoci commented:
@shanaka-rajitha is this design still valid in respect to https://xd.adobe.com/view/0790f950-abbd-41fa-b372-332295fd876f-52c3/?x_product=cc-slack%2F1.5.1 ? If not, can you please update accordingly?

@lo5 I suggest renaming this component to `text_annotator` to allow for multiple types of annotators in future, e.g. `image_annotator`. wdyt?

lo5 commented:
> I suggest renaming this component to text_annotator

Agree

shanaka-rajitha commented:
> is this design still valid in respect to https://xd.adobe.com/view/0790f950-abbd-41fa-b372-332295fd876f-52c3/?x_product=cc-slack%2F1.5.1 ? If not, can you please update accordingly?

@mturoci will update the following component with minor changes to match new design guidelines.

shihan007 commented:
@mturoci https://xd.adobe.com/view/0790f950-abbd-41fa-b372-332295fd876f-52c3/screen/78995835-9119-4777-b4db-b7f5dd0e7bf6
<img width="735" alt="Screenshot 2021-10-13 at 12 16 21" src="https://user-images.githubusercontent.com/38729017/137081573-6caf1c27-4ec9-4037-82e5-1172516065ca.png">

shihan007 commented:
> ![leftright](https://user-images.githubusercontent.com/6334572/116582391-2d315500-a8ca-11eb-8e42-013e146b66d5.png)
> 
> If we go with the right-aligned 'x' button strategy, the 'x' will end up in the lower-middle [1] if the text wraps.

same issue occur if we have highlighted the top text. Btw I am okay with placing it right @mturoci @lo5 

cc: @shanaka-rajitha
